runtime/vm/mz64/UpcallThunkGen.cpp

```diff
@@ -56,179 +57,433 @@ extern "C" {
-                }
-                currArgIndex += 1;
-                fprIndex += 2;
-                if (i < 3) {
-                    gprIndex++; // If one of the first three doublewords of the argument area is float/double, then the corresponding GPR is undefined. So we must ignore any value in there.
-                }
-                break;
-            default:
-                Assert_VM_unreachable();
-        }
+	// Now we loop through the arguments and generate the actual instructions to store any register arguments in the caller frame.
+	const I_32 numRegsToPreserve = 10;
+	const I_32 offsetToRegSaveArea = defaultStackPointerBias - thunkFrameSize + registerSaveAreaSize - (numRegsToPreserve * 8); // This should evaluate to 1808.
+	thunkMem += STMG(thunkMem, R6, R15, R4, offsetToRegSaveArea);
+	thunkMem += LAY(thunkMem, R4, 0, R4, -thunkFrameSize);
+	if (hiddenParameter) {
+		thunkMem += STG(thunkMem, gprIndex, 0, R4, argAreaOffsetFromSP + (currStackSlotIndex * 8));
+	gprIndex++;
+	currStackSlotIndex++;
+	}
+	for (I_32 i = 0; i < lastSigIdx; i++) {
+		U_32 argSizeInBytes = sigArray[i].sizeInByte;
+		switch(sigArray[i].type) {
+		case J9_FFI_UPCALL_SIG_TYPE_CHAR:    /* Fall through */
+		case J9_FFI_UPCALL_SIG_TYPE_SHORT:   /* Fall through */
+		case J9_FFI_UPCALL_SIG_TYPE_INT32:   /* Fall through */
+		case J9_FFI_UPCALL_SIG_TYPE_POINTER: /* Fall through */
+		case J9_FFI_UPCALL_SIG_TYPE_INT64:
+			if (gprIndex <= maxGPRIndex) {
+				thunkMem += STG(thunkMem, gprIndex, 0, R4, argAreaOffsetFromSP + (currStackSlotIndex * 8));
+				gprIndex += 1;
+			}
+			currStackSlotIndex += 1;
+			break;
+		case J9_FFI_UPCALL_SIG_TYPE_FLOAT:
+			if (fprIndex <= maxFPRIndex) {
+				thunkMem += STEY(thunkMem, fprIndex, 0, R4, argAreaOffsetFromSP + (currStackSlotIndex * 8));
+				fprIndex += 2;
+			}
+			currStackSlotIndex += 1;
+			if (i < 3) {
+				gprIndex++; // If one of the first three doublewords of the argument area is float/double, then the corresponding GPR is undefined. So we must ignore any value in there.
+			}
+			break;
+		case J9_FFI_UPCALL_SIG_TYPE_DOUBLE:
+			if (fprIndex <= maxFPRIndex) {
+				thunkMem += STDY(thunkMem, fprIndex, 0, R4, argAreaOffsetFromSP + (currStackSlotIndex * 8));
+				fprIndex += 2;
+			}
+			currStackSlotIndex += 1;
+			if (i < 3) {
+				gprIndex++; // If one of the first three doublewords of the argument area is float/double, then the corresponding GPR is undefined. So we must ignore any value in there.
+			}
+			break;
+		case J9_FFI_UPCALL_SIG_TYPE_STRUCT_AGGREGATE_ALL_SP:
+			if (argSizeInBytes == 2 * sizeof(float)) {
+				// a pure float struct with only 2 floats is passed in 2 FPRs.
+				// If only one FPR is free (i.e. FPR6), then the first half of the struct is
+				// in FPR6, and the second half is in the corresponding slot
+				// in the argument area, right adjusted.
+				if (fprIndex <= maxFPRIndex - 1) {
```

Comment:
Why are we not using `fprIndex < maxFPRIndex - 2)` here ? Wouldn't that be more simpler for readability?

Reply:
Do you mean it's easier for readability because I was subtracting by `2` in the previous loop?

Reply:
This is fixed in 53c52c9835f66d28f5b4b3e0120ae281ed8f0b02